processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -384,55 +379,115 @@ public int add(InputRow row) throws IndexSizeExceededException
       throw new IAE("Cannot add row[%s] because it is below the minTimestamp[%s]", row, new DateTime(minTimestamp));
     }
 
+    String[][] dims = null;
     final List<String> rowDimensions = row.getDimensions();
 
-    String[][] dims;
-    List<String[]> overflow = null;
-    synchronized (dimensionOrder) {
-      dims = new String[dimensionOrder.size()][];
-      for (String dimension : rowDimensions) {
-        List<String> dimensionValues = row.getDimension(dimension);
-
-        // Set column capabilities as data is coming in
-        ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
-        if (capabilities == null) {
-          capabilities = new ColumnCapabilitiesImpl();
-          capabilities.setType(ValueType.STRING);
-          columnCapabilities.put(dimension, capabilities);
+    for (String dimension : rowDimensions) {
+      List<String> dimensionValues = row.getDimension(dimension);
+
+      // Set column capabilities as data is coming in
+      ColumnCapabilitiesImpl capabilities = columnCapabilities.get(dimension);
+      if (capabilities == null) {
+        capabilities = new ColumnCapabilitiesImpl();
+        capabilities.setType(ValueType.STRING);
+        ColumnCapabilitiesImpl other = columnCapabilities.putIfAbsent(dimension, capabilities);
+        if (null != other) {
+          // Lost a race
+          capabilities = other;
         }
-        if (dimensionValues.size() > 1) {
-          capabilities.setHasMultipleValues(true);
-        }
-
-        Integer index = dimensionOrder.get(dimension);
-        if (index == null) {
-          dimensionOrder.put(dimension, dimensionOrder.size());
-          dimensions.add(dimension);
+      }
+      if (dimensionValues.size() > 1) {
+        capabilities.setHasMultipleValues(true);
+      }
 
-          if (overflow == null) {
-            overflow = Lists.newArrayList();
-          }
-          overflow.add(getDimVals(dimValues.add(dimension), dimensionValues));
-        } else {
-          dims[index] = getDimVals(dimValues.get(dimension), dimensionValues);
+      Integer orderID = dimensionOrder.putIfAbsent(dimension, raceChecker);
+      while (orderID != null && orderID < 0) {
```

Comment:
if the other thread dies or throws an exception we are now spinning forever, so we need to make sure the (-1) placeholder lock can get released.